### PR TITLE
Add xml comments to Validation namespace

### PR DESCRIPTION
**Pull request by Shane32**

### CODE REVIEW

File: src/GraphQL.ApiTests/GraphQL.approved.txt [hunk 1]
```diff
@@ -3023,6 +3025,7 @@ namespace GraphQL.Validation.Rules
     }
     public class OverlappingFieldsCanBeMerged : GraphQL.Validation.IValidationRule
     {
+        public static readonly GraphQL.Validation.Rules.OverlappingFieldsCanBeMerged Instance;
```
Comment (sungam3r) [inline src/GraphQL.ApiTests/GraphQL.approved.txt:3028]: 👍 I noticed this unused rule too.

File: src/GraphQL/Validation/BasicVisitor.cs [hunk 2]
```diff
@@ -4,20 +4,31 @@
 
 namespace GraphQL.Validation
 {
+    /// <summary>
+    /// Walks an AST node tree executing <see cref="INodeVisitor.Enter(INode)"/> and <see cref="INodeVisitor.Leave(INode)"/> events for each node.
```
Comment (sungam3r) [inline src/GraphQL/Validation/BasicVisitor.cs:8]:
```suggestion
    /// Walks an AST node tree executing <see cref="INodeVisitor.Enter(INode)"/> and <see cref="INodeVisitor.Leave(INode)"/> methods for each node.
```

File: src/GraphQL/Validation/BasicVisitor.cs [hunk 3]
```diff
@@ -4,20 +4,31 @@
...
 
+        /// <inheritdoc cref="BasicVisitor.BasicVisitor(INodeVisitor[])"/>
         public BasicVisitor(IList<INodeVisitor> visitors)
         {
             _visitors = visitors;
         }
 
+        /// <summary>
+        /// Walks the specified <see cref="INode"/>, executing <see cref="INodeVisitor.Enter(INode)"/> and
+        /// <see cref="INodeVisitor.Leave(INode)"/> events for each node.
```
Comment (sungam3r) [inline src/GraphQL/Validation/BasicVisitor.cs:30]:
```suggestion
        /// <see cref="INodeVisitor.Leave(INode)"/> methods for each node.
```

File: src/GraphQL/Validation/EnterLeaveListener.cs [hunk 4]
```diff
@@ -36,6 +46,12 @@ void INodeVisitor.Leave(INode node)
             }
         }
 
+        /// <summary>
+        /// Configures an event listener for the specified node type.
+        /// </summary>
+        /// <typeparam name="TNode">The type of the AST node to listen for.</typeparam>
+        /// <param name="enter">A delegate to execute when the specified node type is entered.</param>
```
Comment (sungam3r) [inline src/GraphQL/Validation/EnterLeaveListener.cs:53]:
```suggestion
        /// <param name="enter">A delegate to execute when the node of specified type is entered.</param>
```

File: src/GraphQL/Validation/EnterLeaveListener.cs [hunk 5]
```diff
@@ -36,6 +46,12 @@ void INodeVisitor.Leave(INode node)
...
+        /// <param name="leave">A delegate to execute when the specified node type is left.</param>
```
Comment (sungam3r) [inline src/GraphQL/Validation/EnterLeaveListener.cs:54]:
```suggestion
        /// <param name="leave">A delegate to execute when the node of specified type is left.</param>
```

File: src/GraphQL/Validation/INodeVisitor.cs [hunk 6]
```diff
@@ -2,9 +2,19 @@
 
 namespace GraphQL.Validation
 {
+    /// <summary>
+    /// An interface to handle events raised by a node walker such as <see cref="BasicVisitor"/>.
+    /// </summary>
     public interface INodeVisitor
     {
+        /// <summary>
+        /// Raised when the node walker is entering a node.
```
Comment (sungam3r) [inline src/GraphQL/Validation/INodeVisitor.cs:11]:
```suggestion
        /// Called when the node walker is entering a node.
```

File: src/GraphQL/Validation/INodeVisitor.cs [hunk 7]
```diff
@@ -2,9 +2,19 @@
...
+        /// </summary>
         void Enter(INode node);
+
+        /// <summary>
+        /// Raised when the node walker is leaving a node.
```
Comment (sungam3r) [inline src/GraphQL/Validation/INodeVisitor.cs:16]:
```suggestion
        /// Called when the node walker is leaving a node.
```

File: src/GraphQL/Validation/INodeVisitor.cs [hunk 8]
```diff
@@ -2,9 +2,19 @@
 
 namespace GraphQL.Validation
 {
+    /// <summary>
+    /// An interface to handle events raised by a node walker such as <see cref="BasicVisitor"/>.
```
Comment (sungam3r) [inline src/GraphQL/Validation/INodeVisitor.cs:6]: I'm not sure about _events_ wording.

File: src/GraphQL/Validation/MatchingNodeVisitor.cs [hunk 9]
```diff
@@ -3,12 +3,19 @@
 
 namespace GraphQL.Validation
 {
+    /// <summary>
+    /// A node listener which runs configured delegates only when the node enetered/left matches the specified node type.
```
Comment (sungam3r) [inline src/GraphQL/Validation/MatchingNodeVisitor.cs:7]:
```suggestion
    /// A node listener which runs configured delegates only when the node entered/left matches the specified node type.
```

File: src/GraphQL/Validation/INodeVisitor.cs (hunk 8 above)
Comment (Shane32) [reply src/GraphQL/Validation/INodeVisitor.cs:6]: Events, while not technically C# "events", are easy to understand and portray what's going on rather well.  If you want to change it, that's fine, however.

Comment (Shane32) [reply src/GraphQL/Validation/INodeVisitor.cs:6]:
```suggestion
    /// An interface to handle methods called by a node walker such as <see cref="BasicVisitor"/>.
```

File: src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs [hunk 10]
```diff
@@ -8,8 +8,25 @@
...
 {
+    /// <summary>
+    /// Overlapping fields are mergable:
+    ///
+    /// If multiple field selections with the same response names are encountered during execution,
+    /// the field and arguments to execute and the resulting value should be unambiguous. Therefore
+    /// any two field selections which might both be encountered for the same object are only valid
+    /// if they are equivalent.
+    /// <br/><br/>
+    /// Due to the complexity of the implementation, this validation rule is not part of the default set of validation rules.
```
Comment (sungam3r) [inline src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:19]: I thought to add it.

Comment (sungam3r) [reply src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:19]: By the way, this rule has bug.
изображение

File: src/GraphQL/Validation/TypeInfo.cs [hunk 11]
```diff
@@ -5,6 +5,9 @@
 
 namespace GraphQL.Validation
 {
+    /// <summary>
+    /// Provides information pertaining to the current state of the AST tree while being walked.
```
Comment (sungam3r) [inline src/GraphQL/Validation/TypeInfo.cs:9]: This is one of the most confusing classes for me.

File: src/GraphQL/Validation/ValidationContext.cs [hunk 12]
```diff
@@ -58,6 +83,10 @@ public List<VariableUsage> GetVariables(IHaveSelectionSet node)
             return usages;
         }
 
+        /// <summary>
+        /// For a specified operation with a document, returns a list of variable references
+        /// along with what input type each were referenced for.
```
Comment (sungam3r) [inline src/GraphQL/Validation/ValidationContext.cs:88]: _was_ referenced ?

File: src/GraphQL/Validation/ValidationContext.cs [hunk 13]
```diff
@@ -148,22 +186,42 @@ public IEnumerable<FragmentDefinition> GetRecursivelyReferencedFragments(Operati
...
+        /// Returns the name of the specified graph type.
+        /// </summary>
         public string Print(IGraphType type)
         {
             return SchemaPrinter.ResolveName(type);
         }
     }
 
+    /// <summary>
+    /// Represents a variable reference node and the graph type it is referenced to be used for.
```
Comment (sungam3r) [inline src/GraphQL/Validation/ValidationContext.cs:207]: I'm not a native speaker and this comment seems hard to understand.

Comment (Shane32) [reply src/GraphQL/Validation/ValidationContext.cs:207]: Yeah I know...... any ideas? Sometimes I just draw a blank about what to write.

File: src/GraphQL/Validation/DocumentValidator.cs [hunk 14]
```diff
@@ -18,8 +22,12 @@ public interface IDocumentValidator
             Inputs inputs = null);
     }
 
+    /// <inheritdoc/>
     public class DocumentValidator : IDocumentValidator
     {
+        /// <summary>
+        /// Returns the default set of validation rules: all except <see cref="OverlappingFieldsCanBeMerged"/>.
```
Comment (sungam3r) [inline src/GraphQL/Validation/DocumentValidator.cs:29]: How hard is this rule? Is everything so bad with it?

File: src/GraphQL/Validation/ValidationContext.cs (hunk 13 above)
Comment (sungam3r) [reply src/GraphQL/Validation/ValidationContext.cs:207]: OK... let's keep it as is.

File: src/GraphQL/Validation/DocumentValidator.cs (hunk 14 above)
Comment (Shane32) [reply src/GraphQL/Validation/DocumentValidator.cs:29]: I don't know.  But it's probably twice as much code as all the rest of the validation rules combined.  Creates a bunch of lists and stuff.  Take a look.  I figured that for 3.3 we probably wouldn't be changing the core set of rules, and in 4.0 we can decide if that behavior should be changed.

Comment (Shane32) [reply src/GraphQL/Validation/DocumentValidator.cs:29]: But at least it is obvious now that there's a validation rule that is not in use.

File: src/GraphQL/Validation/ValidationContext.cs (hunk 12 above)
Comment (Shane32) [reply src/GraphQL/Validation/ValidationContext.cs:88]:
```suggestion
        /// along with what input type each was referenced for.
```

File: src/GraphQL/Validation/TypeInfo.cs (hunk 11 above)
Comment (Shane32) [reply src/GraphQL/Validation/TypeInfo.cs:9]: Yup, but I expect that it saves a lot of allocations and code.

File: src/GraphQL/Validation/DocumentValidator.cs (hunk 14 above)
Comment (sungam3r) [reply src/GraphQL/Validation/DocumentValidator.cs:29]: Let's make new issue about this rule and keep it as is for now.

File: src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs (hunk 10 above)
Comment (Shane32) [reply src/GraphQL/Validation/Rules/OverlappingFieldsCanBeMerged.cs:19]: I suggest that we add it, or not, in 4.0.

File: src/GraphQL/Validation/IValidationRule.cs [hunk 15]
```diff
@@ -2,8 +2,15 @@
...
 namespace GraphQL.Validation
 {
+    /// <summary>
+    /// Represents a validation rule for a document.
+    /// </summary>
     public interface IValidationRule
     {
+        /// <summary>
+        /// Validates a document against this validation rule. Validation failures are added
+        /// to a list stored within <see cref="ValidationContext.Errors"/>.
```
Comment (Shane32) [inline src/GraphQL/Validation/IValidationRule.cs:12]:
```suggestion
        /// Prepares and returns a node listener to be used to validate a document (via a node walker) against this
        /// validation rule. Validation failures are added to a list stored within <see cref="ValidationContext.Errors"/>.
```

Comment (Shane32) [reply src/GraphQL/Validation/IValidationRule.cs:12]: How's that?

Comment (sungam3r) [reply src/GraphQL/Validation/IValidationRule.cs:12]: Edited.

**Comment by Shane32** [comment]
I added the one line bug you mentioned to this PR.

**Review by sungam3r — APPROVED** [review]

**Pull request merged**